### ISSUE TITLE
Memory leak or memory not being released

### ISSUE THREAD

**Issue by sanjaysrikakulam** [jnwatson/py-lmdb#302]
### Affected Operating Systems
* Linux

### Affected py-lmdb Version

version 1.2.1

### py-lmdb Installation Method

pip install lmdb

### Distribution name and LMDB library version

Use "print lmdb.version()" from a Python prompt.
0, 9, 29

### Other important machine info

CentOS 7
NVMe disk storage
Python version 3.7.7

### Describe Your Problem

```
import numpy as np
import lmdb
import zstd
import msgpack
from numcodecs import PackBits

class LMDBStore:
    """LMDB backed key-value store

    Notes:
      MessagePack is used for serializing keys
      Zstandard is used for compressing the serialized values
      Flush and close methods must be called exclusively to make sure that the writes are synced to the disk

    Args:
      path: Directory path or file name prefix where this DB is (to be) stored
      readonly: Whether to open the LMDB as read only [False]
      lock: Whether to lock LMDB during concurrent writing [True]
      **kwargs: Keyword arguments passed through to the 'lmdb.open' function

    """

    def __init__(self, path, readonly=False, lock=True, **kwargs):
        # Set LMDB defaults and user provided config to 'lmdb.open'
        # Default map size is set to 2TiB (only on 64 bit systems, should be < 2GB for 32 bit systems)
        kwargs.setdefault("map_size", 2199023255552)
        kwargs.setdefault("readonly", readonly)
        kwargs.setdefault("metasync", False)
        kwargs.setdefault("sync", False)

        # Enable writemap only on Linux systems
        # LMDB doc: This option may cause filesystems that don’t support
        # sparse files, such as OSX, to immediately preallocate map_size= bytes
        # of underlying storage when the environment is opened or closed for
        # the first time.
        if sys.platform.startswith("linux"):
            kwargs.setdefault("writemap", True)

        kwargs.setdefault("map_async", False)
        # kwargs.setdefault("max_dbs", len(db_names))
        kwargs.setdefault("max_spare_txns", psutil.cpu_count())
        kwargs.setdefault("lock", lock)

        path = os.path.abspath(path)
        self.env = lmdb.open(path, **kwargs)
    .
    .
    .
    .
    .
    def compress_nparray(self, value):
        """Serialize and Zstandard compress the value

        Args:
          value: 1 D NumPy array

        Returns: Compressed serialized bytes

        """
        return zstd.compress(PackBits().encode(value), 9, 2)

    def serialize(self, item):
        """Serialization using MessagePack

        Args:
          item: Any item (int, str, list, dict) to serialize

        Returns: Serialized object

        """
        return msgpack.dumps(item)

    def batch_set_nparrays(self, np_array, enum_start):
        """Batch insert

        Args:
          np_array: 2D NumPy boolean array

        """
        with self.env.begin(write=True) as txn:
            with txn.cursor() as cursor:
                for key, row in enumerate(np_array, start=enum_start):
                    if row.nonzero()[0].size > 0:
                        cursor.put(self.serialize(key), self.compress_nparray(row))

    def flush(self):
        """Sync/Write data to the file system"""
        self.env.sync()

    def close(self):
        """Close the DB"""
        self.env.close()

def chunks(samples_list, chunk_size):
    """Given a sample list and chunk size, returns a chunk from the list

    Args:
      samples_list: List object
      chunk_size: Number of items to have in each chunk

    Returns: Generator object of chunks of the input list

    """
    for i in range(0, len(samples_list), chunk_size):
        yield samples_list[i : i + chunk_size]

db = LMDBStore("/data/test.lmdb")
for chunk in chunks(range(600_000_000), 600_000):
    np_array = np.empty((len(chunk), 97237), dtype=bool) # This is a simulated dataset here (54.4GiB memory is consumed)
    db.batch_set_nparrays(np_array, chunk.start)
    db.flush()
db.close()
```

### Problem
Expected that no more than 1GiB will be used in addition to the 54.4GiB NumPy array. But in every iteration, it consumes at least 100 MiB extra and in 20 iterations at least 3 to 4GiB of memory and it is increasing steadily. I do not understand this behaviour. I thought when `sync` is called explicitly everything should be written to disk and the memory should be released. When I close and re-open the `db` inside the loop I do not have this problem. Also, I am not sure about the performance cost in doing the re-opening and closing of `db` inside the loop. Any help regarding this would be really great.

### Additional info
1) Memory consumption or increase was monitored in `htop` `RES`
2) I also tried putting `gc.collect()` in different places, and it did not help.

Thanks in advance!

**Comment by jnwatson** [comment]
Please see https://lmdb.readthedocs.io/en/release/#memory-usage.

TLDR: in this case, resident memory is insufficiently precise to give a useful understanding of memory usage. In reality it reported how much the OS has deigned to keep the memory-mapped file in cache.

**Comment by sanjaysrikakulam** [comment]
Thank you, I did look into it before posting, but thought this behaviour must be due to something else. Can you please comment on the performance cost if one does re-open and close the `db` inside a for loop?

**Comment by jnwatson** [comment]
The performance cost is that the entire database is evicted from the application's address space (but not necessarily the OS cache). Lots of TLBs cleared and then set again. This isn't free.
